docker/indexer/storage/storage.go

```diff
@@ -18,22 +18,82 @@ package storage
 
 import (
 	"context"
+	"fmt"
+	"sync"
+	"time"
 
+	"cloud.google.com/go/datastore"
 	"github.com/go-git/go-git/v5/plumbing"
 	"github.com/google/osv.dev/docker/indexer/stages/preparation"
 	"github.com/google/osv.dev/docker/indexer/stages/processing"
+)
 
-	log "github.com/golang/glog"
+const (
+	kind   = "source-repository"
+	keyFmt = "%s-%x"
 )
 
+// document represents a single repository entry in datastore.
+type document struct {
+	Name        string
+	BaseCPE     string
+	Version     string
+	Commit      []byte
+	When        time.Time
+	RepoType    string
+	FileExts    []string
+	FileResults []processing.FileResult
+}
+
+func newDoc(repoInfo *preparation.Result, fileResults []processing.FileResult) *document {
+	return &document{
+		Name:        repoInfo.Name,
+		BaseCPE:     repoInfo.BaseCPE,
+		Version:     repoInfo.Version,
+		Commit:      repoInfo.Commit[:],
+		When:        repoInfo.When,
+		RepoType:    repoInfo.Type,
+		FileExts:    repoInfo.FileExts,
+		FileResults: fileResults,
+	}
+}
+
+// Store provides the functionality to check for existing documents
+// in datastore and add new ones.
 type Store struct {
+	dsCl  *datastore.Client
+	cache sync.Map
+}
+
+// New returns a new Store.
+func New(ctx context.Context, projectID string) (*Store, error) {
+	client, err := datastore.NewClient(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+	return &Store{dsCl: client, cache: sync.Map{}}, nil
 }
 
+// Exists checks whether a name/hash pair already exists in datastore.
 func (s *Store) Exists(ctx context.Context, name string, hash plumbing.Hash) (bool, error) {
-	return false, nil
+	if _, ok := s.cache.Load(fmt.Sprintf(keyFmt, name, hash)); ok {
```

Comment:
Can we key based on the repo URL instead of name? 

We want to eventually run this on all git repos for C/C++ projects that we see in OSV.